ERCS/erc-7715.md

```diff
@@ -0,0 +1,466 @@
+discussions-to: https://ethereum-magicians.org/t/erc-7715-request-permissions-from-wallets/20100
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-05-24
+requires: 4337, 5792, 7679, 7710
+---
+
+## Abstract
+
+We define a new JSON-RPC method `wallet_issuePermissions` for DApp to request permissions from a wallet in order to execute transactions on users’s behalf.  This enables two use cases:
+
+- Executing transactions for users without a wallet connection.
+- Executing transactions for users with a wallet connection that is scoped with permissions.
+
+##Motivation
+
+Currently most DApps implement a flow similar to the following:
+
+```mermaid
+sequenceDiagram
+    User ->>+ DApp: Establish Connection
+    loop For each call
+    DApp ->> User: Propose Call <br> (Transaction)
+    User ->> DApp: Approve Call
+    end 
+```
+
+Each interaction requires the user to sign a transaction with their wallet.  The problems are:
+
+- It can get tedious for the user to manually approve every transaction, especially in highly-interactive applications such as games.
+- It’s impossible to send transactions for users without an active wallet connection.  This invalidates use cases such as subscriptions, passive investments, limit orders, and more.
+
+In the context of AA, there are multiple vendor-specific implementations of session keys, which are temporary keys imbued with specific permissions.  However, since the implementations are vendor-specific, it’s impossible for DApps to “request” session keys from wallets in a unified way, regardless of the specific wallet implementations.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### `wallet_issuePermissions`
+
+We introduce a `wallet_issuePermissions` method for the DApp to request permissions from the wallet.  We’d prefer `wallet_requestPermissions` but it’s already taken with [ERC-2255](./eip-2255.md).
+
+#### Request Specification
+
+```tsx
+type IssuePermissionsRequestParams = {
+  signer?: {
+    type: string
+    data: any
+  }
+ 
+  permissions: {
+    type: string
+    data: any
+    required: boolean
+  }[]
+  
+  expiry: number,
+}
+```
+
+`signer` is an optional field that identifies the key or account associated with the session.  When not specified, it’s implied that the wallet will manage the session.  See the “Wallet-managed Sessions” section for details.
+
+`permissions` is a list of permissions.  See “Permission Types” section for details.
+
+Each permission can be either `required` or not.  If a permission is `required`, the wallet MUST grant the permission if it approves the request.  Otherwise, the wallet may reject the permission even if it approves the request (presumably because it grants other permissions).
```

Comment:
Why isn't everything required? Adding this will make it so hard for a wallet to correctly convey to its user what they are really permitting.

Seems odd for an end-user to optionally set a limit, only to have it ignored. What is the use case for that?

Reply:
When DApp requests permissions, required ones MUST be issued otherwise some flow DApp side will not work. If the permission is not required, user may or may not grant it and the flows dapp side will not break.